### PR TITLE
Graft of 'serial RX and telemetry on same port' feature from Betaflight.

### PR DESCRIPTION
**Pull request by mikeller**
This PR enables sharing of one serial port between serial RX and telemetry. If a port is shared, the speed and inversion settings of the port are dictated by the serial RX protocol used, and any settings for the telemetry protocol are ignored. As a consequence, this will only work if the RX can receive telemetry data at the same speed as the serial RX data it is set to output.

This has only been tested (and found to work fine) with a Naze32 rev 6 as the FC, and an OrangeRx openLRS RX. The combination of SUMD RX and FrSky telemetry over the same serial port works fine with this hardware.

Any feedback on other combinations working / not working is appreciated, and will be integrated.

### CODE REVIEW

**Comment by hydra** [comment]
Interesting.  It's a shame no other receivers will work at the same baud rates for serial rx and telemetry.  I wish the manufacturers would use standard baud rates with no inversion for serialrx and telemetry.  It would make things much easier for everyone.

Please can you fix the unit tests and the CC3D build?

**Comment by mikeller** [comment]
Yes. Ideally the protocol for serial RX and telemetry should be rolled into one anyway, like Jeti EX bus does.

With openLRS hardware, the situation is even such that only one serial port is available and supported by the firmware, so running both things over a single serial port on the flight controller side is probably the only way to get them working reliably.

Sorry for breaking thing, I wasn't aware that they were. I ran `fake_travis_build.sh` before pushing, and that one seemed to go through all fine. Will fix.

**Comment by mikeller** [comment]
@hydra: I just saw that CC3D is broken due to image size. Any preference on what to cull from it, or what do you propose we do?

**Comment by maxsymba** [comment]
Can I use both SBUS and S.Port from X4RSB on same UART (SPRacing, UART3)?

**Comment by mikeller** [comment]
@maxsymba: No. S.Port uses bidirectional communication for telemetry, so it's not possible to combine it with serial RX on the same port, as serial RX and the TX -> RX part of S.Port would interfere.

**Comment by maxsymba** [comment]
@mikeller Oh.. it's bad((.
I have a problem with "no data" for S.Port (or MSP for OSD) on UART1 (USB interface) and meanwhile try to solve it in other way...

**Comment by mikeller** [comment]
@maxsymba: Have you tried if you get data when you are arming the craft (TAKE OFF THOSE PROPS FIRST! :-))? The way it works when sharing a port between MSP (USB, for configurator) and telemetry is that the port does MSP while unarmed, and telemetry while armed. (Also, this means you can't share between OSD and telemetry in a meaningful way.)

**Comment by maxsymba** [comment]
@mikeller The problem is that earlier my OSD worked through MSP at UART1. Then, at some point it has ceased to work. Then, after program manipulations has again earned. And now, again doesn't work with a new version of BF. Now it is working only when USB cable is plugged and connected to configurator(((

**Comment by mikeller** [comment]
@maxsymba: What are your serial port settings? (The `serial ...` lines in a CLI diff / dump?)

**Comment by maxsymba** [comment]
@mikeller 
serial 0 1 115200 57600 0 115200
serial 1 1 115200 57600 0 115200
serial 2 64 115200 57600 0 115200

**Comment by mikeller** [comment]
@maxsymba: That looks all ok and I can't see any conflicts. There have been some changes in the format of the MSP protocol that the latest versions of Betaflight use. It's possible that your OSD board does not work properly because of these protocol changes.

**Comment by maxsymba** [comment]
@mikeller Works fine on UART2. 
So I don't think that there is a protocol mismatch. 
Such impression that the port is simply not active for some reason. 
But OSD on UART1 begin to work at once when the connection to a configurator is established.

**Comment by mikeller** [comment]
This should be good to go now I think.

**Comment by hydra** [comment]
unrelated support comments deleted.

**Comment by hydra** [comment]
seems OK, initially I was confused but then I found the defintion for `TELEMETRY_SHAREABLE_PORT_FUNCTIONS_MASK` and then I was happy. :smile:

**Comment by hydra** [comment]
User documentation is missing for this, please can you create a new PR that adds it?

**Comment by hydra** [comment]
This needs rebasing against master.   

Also I think `telemetryIsPortSharedWithRx` is better name for 'telemetryCheckRxPortShared'.  Can you also fix that?

**Comment by mikeller** [comment]
:-(

Now I get `region `FLASH' overflowed by 466 bytes` for NAZE. Do you want to postpone this, or is there another feature that can be culled?

**Comment by ledvinap** [comment]
That's strange .. this does not look like 450 bytes of code ...
Maybe comparing `.map` files will reveal other problem

**Comment by mikeller** [comment]
@ledvinap I couldn't spot anything untowards, and looking at the changes it _might_ be 450 bytes worth of.
For now I have disabled SONAR on NAZE to make it build (we already had to do this for CC3D_OPBL).

**Comment by martinbudden** [comment]
@hydra , can we merge this? As well as being useful in itself, it is also helps convergence with betaflight: the `rx` and `telemetry` directories are certainly areas where the code should be common.

File: docs/Serial.md [hunk 1]
```diff
@@ -46,8 +46,9 @@ e.g. after configuring a port for GPS enable the GPS feature.
 * If SoftSerial is used, then all SoftSerial ports must use the same baudrate.
 * Softserial is limited to 19200 baud.
 * All telemetry systems except MSP will ignore any attempts to override the baudrate.
-* MSP/CLI can be shared with EITHER Blackbox OR telemetry.  In shared mode blackbox or telemetry will be output only when armed.
+* MSP/CLI can be shared with EITHER Blackbox OR telemetry. In shared mode blackbox or telemetry will be output only when armed.
 * Smartport telemetry cannot be shared with MSP.
+* FrSky and LTM telemetry can be shared with serial RX. In this case, the serial port's RX pin on the flight controller will be used for serial RX, the TX pin for telemetry output. When used this way, the baudrate used for telemetry will be set to the baudrate that is used by serial RX, make sure to check your RX supports this (currently only confirmed working with openLRSng modules).
```
Comment (mikeller) [inline docs/Serial.md]: That's why only FrSky and LTM are supported...

**Comment by unitware** [comment]
Yeah, it was just a comment, and an opportunity to improve the docs if you are inclined to do so. I did some measurements last night to find out if this was the case. 

Of course if the PR is merged before the IBus one - then I can add that myself in the docs.

//M

**Comment by mikeller** [comment]
I'm not sure if the docs need updating. I decided to list the supported protocols (as opposed to the ones not supported) for port sharing. This way, without any protocol being added that explicitly implements sharing, the docs should not need updating.

My hope is that this should be clear enough. If too many user questions arise, I might have to revert to listing both supported and non-supported protocols explicitly.

**Comment by unitware** [comment]
Im fine with that
//M

**Comment by fishpepper** [comment]
> Interesting. It's a shame no other receivers will work at the same baud rates for serial rx and telemetry. I wish the manufacturers would use standard baud rates with no inversion for serialrx and telemetry. It would make things much easier for everyone.
When you flash the frsky receivers with my opensource code ([OpenSky](http://fishpepper.de/projects/opensky/)) you can use the same port for telemetry as you do for sbus at 100000baud ;)
This is a really nice feature, it also allows rx firmware upgrades by using the serial passthrough mode.

File: docs/Serial.md (hunk 1 above)
Comment (fishpepper) [inline docs/Serial.md]: > confirmed working with openLRSng modules)

And frsky or diy receivers running the [OpenSky](http://fishpepper.de/projects/opensky/) firmware ;)

**Comment by unitware** [comment]
I'm testing to combine flysky ibus telemetry and serial rx here, borrowed some snippets in the isSerialConfigurationValid from here.

https://github.com/cleanflight/cleanflight/issues/2545

**Pull request closed without merging**